qualtran/resource_counting/bloq_counts.py

```diff
@@ -69,66 +69,121 @@ def _build_cbloq_counts_graph(cbloq: CompositeBloq) -> Set[BloqCountT]:
-    if keep(parent):
-        return {parent: 1}
+    if keep(bloq):
+        raise DecomposeTypeError("Requested via `keep`.")
 
     # Base case 2: Max depth exceeded
     if max_depth is not None and depth >= max_depth:
-        return {parent: 1}
+        raise DecomposeTypeError("Max depth exceeded.")
 
-    try:
-        count_decomp = parent.build_call_graph(ssa)
-    except (DecomposeNotImplementedError, DecomposeTypeError):
-        # Base case 3: Decomposition (or `bloq_counts`) is not implemented. This is left as a
-        #              leaf node.
-        return {parent: 1}
+    # Prep for recursion: get the callees and modify them according to `generalizer`.
+    callee_counts = _generalize_callees(bloq.build_call_graph(ssa), generalizer)
 
-    sigma: Dict[Bloq, Union[int, sympy.Expr]] = defaultdict(lambda: 0)
-    for child, n in count_decomp:
-        child = generalizer(child)
-        if child is None:
-            continue
+    # Base case 3: Empty list of callees
+    if not callee_counts:
+        raise DecomposeTypeError("Empty list of callees.")
+
+    for callee, n in callee_counts:
+        # Quite important: we do the recursive call first before adding in the edges.
+        # Otherwise, adding the edge would mark the callee node as already-visited by
+        # virtue of it being added to the graph with the `g.add_edge` call.
+
+        # Do the recursive step, which will continue to mutate `g`
+        _build_call_graph(callee, generalizer, ssa, keep, max_depth, g, depth + 1)
 
         # Update edge in `g`
-        if (parent, child) in g.edges:
-            g.edges[parent, child]['n'] += n
+        if (bloq, callee) in g.edges:
+            g.edges[bloq, callee]['n'] += n
         else:
-            g.add_edge(parent, child, n=n)
+            g.add_edge(bloq, callee, n=n)
 
-        # Do the recursive step, which will continue to mutate `g`
-        child_counts = _recurse_call_graph(child, g, ssa, generalizer, keep, max_depth, depth + 1)
 
-        # Update `sigma` with the recursion results.
-        for k in child_counts.keys():
-            sigma[k] += child_counts[k] * n
+def _build_call_graph(
+    bloq: Bloq,
+    generalizer: Callable[[Bloq], Optional[Bloq]],
+    ssa: SympySymbolAllocator,
+    keep: Callable[[Bloq], bool],
+    max_depth: Optional[int],
+    g: nx.DiGraph,
+    depth: int,
+) -> None:
+    """Recursively build the call graph.
+
+    Arguments are the same as `get_bloq_call_graph`, except `g` is the graph we're building
+    (i.e. it is mutated by this function) and `depth` is the current recursion depth.
+    """
+    if bloq in g:
+        # We already visited this node.
+        return
+
+    # Make sure this node is present in the graph. You could annotate
+    # additional node properties here, too.
+    g.add_node(bloq)
 
+    try:
+        return _build_call_graph_inner(bloq, generalizer, ssa, keep, max_depth, g, depth)
```

Comment:
nit: If I were you, I'd keep the logic of ` _build_call_graph_inner` directly here instead of abstracting away in a separate function. The reason is that I'm not a fan of `mutual recursion` (i.e. a->b->a->b) unless it's absolutely necessary. 


I see the extra level of indentation in the `try: / except:` bloq can be annoying, but to get rid of that I'd rather have a helper method `_get_callee_counts(*args)` and return an empty dict when `keep(bloq)` or `(max_depth is not None and depth >= max_depth)` is satisfied. So, 

```python
def _get_callee_counts(*args):
     if keep(bloq) or (max_depth is not None and depth >= max_depth):
        # Base cases 1 & 2 when `bloq` is a leaf.
        return {}
     return _generalize_callees(bloq.build_call_graph(ssa), generalizer)

def _build_call_graph(*args):
    callee_counts = _get_callee_counts(*args)
    if not callee_counts:
        # We don't do anything bespoke for leaf nodes, but we could...
        return 
    for callee, n in callee_counts:
        # continue with normal logic here
```

You could also raise either of `(DecomposeNotImplementedError, DecomposeTypeError)` in `_get_callee_counts` and catch it in `_build_call_graph` but instead of checking for `if not callee_counts` if you want. But loop that iterates on the children and recursively calls `_build_call_graph` would still be part of the `_build_call_graph` implementation.

Reply:
done